worker/peergrouper/desired.go

```diff
@@ -192,6 +201,57 @@ func desiredPeerGroup(info *peerGroupInfo) (map[string]*replicaset.Member, map[s
 	}
 	return newMembers, machineVoting, nil
 }
+// reviewPeerGroupChanges adds some extra logic after creating possiblePeerGroupChanges to safely add
+// or remove machines keeping the correct odd number of voters peer structure, and preventing the
+// primary peer demoting.
+func reviewPeerGroupChanges(newMembers map[string]*replicaset.Member,
+							info *peerGroupInfo,
+							peerChanges *peerGroupChanges){
+
+	currVoters := 0
+	for _, m := range newMembers {
+		if isVotingMember(m) {
+			currVoters += 1
+		}
+	}
+	keptVoters := currVoters - len(peerChanges.toRemoveVote)
+	if (keptVoters+len(peerChanges.toAddVote))%2 != 1 {
```

Comment:
I think we can make the code clearer since we now have a separate function.
If we do:

if (keptVoters +len(peerChanges.toAddVote) %2 == 1 {
  return // we have an odd number of desired voters
}